pkg/kubemark/cloudprovider.go

```diff
@@ -0,0 +1,368 @@
+	currSize, err := kubemarkProvider.GetNodeGroupSize(nodeGroup)
+	if err != nil {
+		return err
+	}
+	switch delta := size - currSize; {
+	case delta < 0:
+		return kubemarkProvider.removeNodesFromNodeGroup(nodeGroup, -delta)
+	case delta > 0:
+		return kubemarkProvider.addNodesToNodeGroup(nodeGroup, delta)
+	}
+	return nil
+}
+
+func (kubemarkProvider *Provider) addNodesToNodeGroup(nodeGroup string, delta int) error {
+	if delta < 0 {
+		return fmt.Errorf("delta has to be positive. Got %d", delta)
+	}
+	for i := 0; i < delta; i++ {
+		if err := kubemarkProvider.addNodeToNodeGroup(nodeGroup); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (kubemarkProvider *Provider) removeNodesFromNodeGroup(nodeGroup string, delta int) error {
+	if delta < 0 {
+		return fmt.Errorf("delta has to be positive. Got %d", delta)
+	}
+	nodes, err := kubemarkProvider.GetNodesForNodegroup(nodeGroup)
+	if err != nil {
+		return err
+	}
+	for i, node := range nodes {
+		if i == delta {
+			return nil
+		}
+		if err := kubemarkProvider.removeNodeFromNodeGroup(nodeGroup, node); err != nil {
+			return err
+		}
+	}
+	return fmt.Errorf("can't remove %d nodes from %s nodegroup, not enough nodes", delta, nodeGroup)
+}
+
+func (kubemarkProvider *Provider) removeNodeFromNodeGroup(nodeGroup string, node string) error {
+	for _, podObj := range kubemarkProvider.externalCluster.podLister.GetStore().List() {
+		pod, ok := podObj.(*apiv1.Pod)
+		if !ok {
+			continue
+		}
+		if pod.ObjectMeta.Name == node {
+			if pod.ObjectMeta.Labels[nodeGroupLabel] != nodeGroup {
+				return fmt.Errorf("can't delete node %s from nodegroup %s. Node is not in nodegroup", node, nodeGroup)
+			}
+			policy := metav1.DeletePropagationForeground
+			err := kubemarkProvider.externalCluster.client.CoreV1().ReplicationControllers(namespaceKubemark).Delete(
+				pod.ObjectMeta.Labels["name"],
+				&metav1.DeleteOptions{PropagationPolicy: &policy})
+			if err != nil {
+				return err
+			}
+			glog.Infof("marking node %s for deletion", node)
+			// Mark node for deleteion from kubemark cluster.
+			// Once it becomes unready after replication controller
+			// deletion has been noticed, we will delete it explicitly.
+			// This is to cover for the fact that kubemark does not
+			// take care of this itself.
+			kubemarkProvider.kubemarkCluster.markNodeForDeletion(node)
+			return nil
+		}
+	}
+
+	return fmt.Errorf("can't delete node %s from nodegroup %s. Node does not exist", node, nodeGroup)
+}
+
+func (kubemarkProvider *Provider) getNodeGroupByName(nodeGroup string) *apiv1.ReplicationController {
+	for _, obj := range kubemarkProvider.externalCluster.rcLister.GetStore().List() {
+		rc, ok := obj.(*apiv1.ReplicationController)
+		if !ok {
+			continue
+		}
+		if rc.ObjectMeta.Labels[nodeGroupLabel] == nodeGroup {
+			return rc
+		}
+	}
+	return nil
+}
+
+func (kubemarkProvider *Provider) addNodeToNodeGroup(nodeGroup string) error {
```

Comment:
nit: Could you move this function to below addNodesToNodeGroup just for better readability.